File: CS102 Prep/Easy - TwoSum.java

import java.util.HashMap;

/*
++++++++++++++++++++++++++++
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
Example 2:

Input: nums = [3,2,4], target = 6
Output: [1,2]
Example 3:

Input: nums = [3,3], target = 6
Output: [0,1]
 

Constraints:

2 <= nums.length <= 104
-109 <= nums[i] <= 109
-109 <= target <= 109
Only one valid answer exists.
 

Follow-up: Can you come up with an algorithm that is less than O(n2) time complexity?
++++++++++++++++++++++++++++
Breakdown:
1. O(n2) -> Use of the difference, nest loop to check the subsequent numbers are the same as the difference. Else, nothing is returned.
2. O(n) -> Use hashmap (index of number: difference from target), compute the difference.Store the difference in hashmap. For every number, check if the difference number exists in hashmap. returns the current index, and the index of the number with the same difference.
++++++++++++++++++++++++++++
Key Takeaways:
- Arrays.length to call upon. 
 */

 class Solution { 
    
    // O(n2) solution
    public int[] twoSum(int[] nums, int target) {
        
        for (int i=0; i < nums.length;i++){
            int diff = target - nums[i];

            for (int x = i+1; x < nums.length; x++){
                if (diff ==  nums[x]){
                    return new int[]{i, x};
                }
            }
        }
        return new int[]{};
    }

    // O(n)
    // Uses hashmap to map the complement, based on the concept of using the hash collision.
    public int[] twoSum_On(int[] nums, int target) {
        HashMap<Integer,Integer> mapList = new HashMap<>();
        Integer value = null;

        for (int i = 0; i < nums.length; i++) { //Iterating -> O(n)
            int pointerNumb = nums[i]; //searching - indexing -> O(1)
            int complement = target - pointerNumb; 
            
            if (i == 0){
               value = mapList.get(complement); //indexing for hashmap -> O(1) 
            }
            else{
                value = mapList.get(pointerNumb);
            }
            
            if (value != null){
                return new int[]{i, value};
            }else{
                mapList.put(complement, i);
            }

        }
        return new int[]{};
    }


}

File: python/Easy - TwoSum.py
"""
++++++++++++++++++++++++++++
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

 
Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
Example 2:

Input: nums = [3,2,4], target = 6
Output: [1,2]
Example 3:

Input: nums = [3,3], target = 6
Output: [0,1]
++++++++++++++++++++++++++++
"""
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        diffDict = {}
        for curr_index, curr_number in enumerate(nums):
            diff = target - curr_number
            if curr_number in diffDict:
                return [diffDict[curr_number],curr_index]
            diffDict[diff] = curr_index
                

# Today's interesting learning was guard clause cases (reduce the mental strain on the increased indentation) + the use of enumerate in python.
#
# Also hit on an issue while coding this out - an unhashable type. So turns out that python hashes any data that is trying to do a search/comparison.
# But only immutable types in python are hashable, which make sense - like if a List was first hashed, and then the user adds in one more value, the hash of the entire List is going to change.
# So, how do mutable types get hashed?
#

        
        
